edx_proctoring/api.py

```diff
@@ -1623,7 +1613,12 @@ def _get_timed_exam_view(exam, context, exam_id, user_id, course_id):
         # check if the exam's due_date has passed. If so, return None
         # so that the user can see their exam answers in read only mode.
         if not exam['hide_after_due'] and is_exam_passed_due(exam, user=user_id):
-            return None
+            if context.get('grace_period', False):
+                has_context_updated = verify_and_add_wait_deadline(context, exam['due_date'])
```

Comment:
`exam['due_date']` won't be accurate, since it doesn't take into account due date extensions. You will have to call `get_exam_due_date(exam, user)` to get the right date.

I recommend refactoring `verify_and_add_wait_deadline()` so that it takes the context, exam, and user_id, and move `get_exam_due_date()` to utils.py. 

Instead of checking for the existence of grace_period here, check it in verify_and_add_wait_deadline.

Reply:
I will utilize the method you mentioned to cater to the individual extensions. As for the grace period check, I wrapped it around the function call to avoid unnecessarily call if the grace period is not found. But I think doing so will help avoid duplication.